### DISCUSSION TITLE
Why AmbientUnitOfWork and UnitOfWorkManager are Singleton

### DISCUSSION THREAD

**Discussion by TomatoDinosaurEgg** [abpframework/abp#12238] [Q&A] [answered]
I try to read source code recently, I fond that "AmbientUnitOfWork" and "UnitOfWorkManager" are Singletons (implement ISingletonDependency), assume all request use the default reservation name is "_AbpActionUnitOfWork", then does that means all the request will use ChildUnitOfWork instance with the same one parent UnitOfWork instance? If so, I see the method CompleteAsync in ChildUnitOfWork just do nothing (return Task.CompletedTask), that really confused me, does anyone know how it works? Thanks

**Comment by maliming** [comment, marked as answer]
hi

**AsyncLocal** is used internally, so there are no conflicts, and since you've looked at the source, I suggest you keep looking at it.

**Reply by TomatoDinosaurEgg** [reply]
hi

Seems I misunderstood with "AsyncLocal", yes you are right, now everything make sense to me, thank you!
